### PR TITLE
feat: Support loading models stored on GCS

### PR DESCRIPTION
**Pull request by dstnluong-google**
# What does this PR do?
If the model_id is a GCS uri, the model will be downloaded and then treated as a local model by changing the model_id to the local path.

This makes google-cloud-storage a new dependency.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by OlivierDehaene** [comment]
@julien-c what do you think?

**Comment by dstnluong-google** [comment]
@OlivierDehaene, @julien-c pinging to ask if there an update for this PR :)

File: launcher/src/main.rs [hunk 1]
```diff
@@ -907,6 +907,9 @@ fn download_convert_model(args: &Args, running: Arc<AtomicBool>) -> Result<(), L
         }
         sleep(Duration::from_millis(100));
     }
+    if args.model_id.starts_with("gs://") {
+        args.model_id = "/tmp/gcs_model/".to_string();
+    }
```
Comment (philschmid) [inline launcher/src/main.rs:912]: We probably want to wraps this for the `google` feature

File: server/requirements_common.txt [hunk 2]
```diff
@@ -8,6 +8,7 @@ einops==0.6.1 ; python_version >= "3.9" and python_version < "3.13"
 filelock==3.13.1 ; python_version >= "3.9" and python_version < "3.13"
 fsspec==2023.10.0 ; python_version >= "3.9" and python_version < "3.13"
 googleapis-common-protos==1.61.0 ; python_version >= "3.9" and python_version < "3.13"
+google-cloud-storage==2.14.0 ; python_version >= "3.9" and python_version < "3.13"
```
Comment (philschmid) [inline server/requirements_common.txt:11]: We probably want to wraps this for the `google` feature

File: server/requirements_cuda.txt [hunk 3]
```diff
@@ -9,6 +9,7 @@ einops==0.6.1 ; python_version >= "3.9" and python_version < "3.13"
 filelock==3.13.1 ; python_version >= "3.9" and python_version < "3.13"
 fsspec==2023.10.0 ; python_version >= "3.9" and python_version < "3.13"
 googleapis-common-protos==1.62.0 ; python_version >= "3.9" and python_version < "3.13"
+google-cloud-storage==2.14.0 ; python_version >= "3.9" and python_version < "3.13"
```
Comment (philschmid) [inline server/requirements_cuda.txt:12]: We probably want to wraps this for the `google` feature

File: server/requirements_rocm.txt [hunk 4]
```diff
@@ -8,6 +8,7 @@ einops==0.6.1 ; python_version >= "3.9" and python_version < "3.13"
 filelock==3.13.1 ; python_version >= "3.9" and python_version < "3.13"
 fsspec==2023.10.0 ; python_version >= "3.9" and python_version < "3.13"
 googleapis-common-protos==1.62.0 ; python_version >= "3.9" and python_version < "3.13"
+google-cloud-storage==2.14.0 ; python_version >= "3.9" and python_version < "3.13"
```
Comment (philschmid) [inline server/requirements_rocm.txt:11]: We probably want to wraps this for the `google` feature

File: server/text_generation_server/cli.py [hunk 5]
```diff
@@ -123,6 +123,10 @@ def download_weights(
     # Import here after the logger is added to log potential import exceptions
     from text_generation_server import utils
 
+    if model_id.startswith(utils.GCS_PREFIX):
+        utils.download_gcs_dir_to_local(model_id, utils.GCS_LOCAL_DIR)
+        model_id = utils.GCS_LOCAL_DIR
```
Comment (philschmid) [inline server/text_generation_server/cli.py:128]: we should make this part of the `launcher` and not load models in python.

File: server/text_generation_server/utils/gcs.py
Comment (philschmid) [inline server/text_generation_server/utils/gcs.py:1]: We should mot make this part of python.

**Review by philschmid — COMMENTED** [review]
Left some comments for @OlivierDehaene and @drbh

**Comment by drbh** [comment]
Hi @dstnluong-google, after some internal discussion we do not think the best path forward is adding/coupling the new dependency. 

Closing the PR for now but am looking forward to future solutions and happy to discuss more in this thread! Thanks you!

**Pull request closed without merging**